README.md
[![Build Status](https://travis-ci.org/tokuhirom/Data-NestedParams.png?branch=master)](https://travis-ci.org/tokuhirom/Data-NestedParams)
# NAME

Data::NestedParams - entry\[title\]=foo&tags\[\]=art&tags\[\]=modern

# SYNOPSIS

    use Data::NestedParams;

    my $expanded = expand_nested_params(
        [
            'entry[title]' => 'foo',
            'tags[]' => 'art',
            'tags[]' => 'modern',
        ]
    );
    # $expanded = { entry => {title => 'foo'}, tags => ['art', 'modern'] };

# DESCRIPTION

Ruby on Rails has a nice feature to create nested parameters that help with the organization of data in a form - parameters can be an arbitrarily deep nested structure.

The way this structure is denoted is that when you construct a form the field names have a special syntax which is parsed.

# TODO

Support `collapse_nested_params()`.

# LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Tokuhiro Matsuno <tokuhirom@gmail.com>

# SEE ALSO

[Catalyst::Plugin::Params::Nested](https://metacpan.org/pod/Catalyst::Plugin::Params::Nested), [CGI::Expand](https://metacpan.org/pod/CGI::Expand)
[https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L90](https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L90)


Changes
Revision history for Perl extension Data-NestedParams

{{$NEXT}}

0.08 2015-11-04T13:00:00Z

    - Collapse now works with JSON boolean values.
      (plainblackguy)

0.07 2014-07-27T23:09:34Z

    - Allow numeric characters in hash key.
      (tokuhirom)

0.06 2014-03-26T11:18:51Z

    - Allow _ and - in hash key.
      (tokuhirom)

0.05 2014-03-21T03:03:24Z

    - Added collapse_nested_query

0.04 2014-03-20T01:24:45Z

    - More rack compatiblity.
    - Documentation.

0.03 2014-03-13T07:49:03Z

    - s/convert_nested_query/expand_nested_params/g

0.02 2014-03-13T07:43:20Z

    - Rename to NestedParams.

0.01 2014-03-13T07:29:00Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Data-NestedParams',
    module_name     => 'Data::NestedParams',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Data/NestedQuery.pm
package Data::NestedQuery;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";

use parent qw(Exporter);

our @EXPORT = qw(convert_nested_query);

# https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L90

sub parse {
    my $k = shift;

    my @keys;
    while (1) {
        if ($k =~ s/\[(\w+)\]\z//) {
            unshift @keys, ['%', $1];
        } elsif ($k =~ s/\[\]\z//) {
            unshift @keys, ['@'];
        } else {
            unshift @keys, ['$', $k];
            last;
        }
    }
    return @keys;
}

sub set_value {
    my ($ret, $k, $v) = @_;

    my @keys = parse($k);

    my $r = \$ret;
    while (@keys) {
        my $key = shift @keys;
        if ($key->[0] eq '%') {
            if (@keys) {
                $r = \(${$r}->{$key->[1]});
            } else { # last
                ${$r}->{$key->[1]} = $v;
            }
        } elsif ($key->[0] eq '@') {
            if (@keys) {
                $r = \(${$r}->[$key->[1]]);
            } else { # last
                push @{$$r}, $v;
            }
        } elsif ($key->[0] eq '$') {
            if (@keys) {
                $r = \(${$r}->{$key->[1]});
            } else {
                ${$r}->{$key->[1]} = $v;
            }
        } else {
            die "ABORT: $key->[0]";
        }
    }
}

sub convert_nested_query {
    my $ary = shift;
    my $ret = +{};
    while (my ($k, $v) = splice @$ary, 0, 2) {
        set_value($ret, $k, $v);
    }
    return $ret;
}

1;
__END__

=encoding utf-8

=head1 NAME

Data::NestedQuery - It's new $module

=head1 SYNOPSIS

    use Data::NestedQuery;

=head1 DESCRIPTION

Data::NestedQuery is ...

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom@gmail.comE<gt>

=cut



lib/Data/NestedParams.pm
package Data::NestedParams;
use 5.008005;
use strict;
use warnings FATAL => 'all';

our $VERSION = "0.08";

use parent qw(Exporter);

our @EXPORT = qw(expand_nested_params collapse_nested_params);

# https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L90
# 9a74ba3b04f2dabe4741d2d82eae723d440c3aa2

sub parse {
    my $k = shift;

    my @keys;
    while (1) {
        if ($k =~ s/\[([a-zA-Z0-9_-]+)\]\z//) {
            unshift @keys, ['%', $1];
        } elsif ($k =~ s/\[\]\z//) {
            unshift @keys, ['@'];
        } else {
            unshift @keys, ['$', $k];
            last;
        }
    }
    return @keys;
}

sub set_value {
    my ($ret, $k, $v) = @_;

    my @keys = parse($k);

    my $r = \$ret;
    while (@keys) {
        my $key = shift @keys;
        if ($key->[0] eq '%') {
            if (@keys) {
                $r = \(${$r}->{$key->[1]});
            } else { # last
                ${$r}->{$key->[1]} = $v;
            }
        } elsif ($key->[0] eq '@') {
            if (@keys) {
                if (defined($$r)) {
                    if (ref($$r->[@$$r-1]) eq 'HASH' && $keys[0]->[0] eq '%' && not exists $$r->[@$$r-1]->{$keys[0]->[1]}) {
                        $r = \(${$r}->[@$$r-1]);
                    } else {
                        $r = \(${$r}->[0+@$$r]);
                    }
                } else {
                    $r = \(${$r}->[0]);
                }
            } else { # last
                push @{$$r}, $v;
            }
        } elsif ($key->[0] eq '$') {
            if (@keys) {
                $r = \(${$r}->{$key->[1]});
            } else {
                ${$r}->{$key->[1]} = $v;
            }
        } else {
            die "ABORT: $key->[0]";
        }
    }
}

sub expand_nested_params {
    my $ary = shift;
    my $ret = +{};
    while (my ($k, $v) = splice @$ary, 0, 2) {
        set_value($ret, $k, $v);
    }
    return $ret;
}

our $COLLAPSE_KEY;

sub _collapse {
    my ($v, $r) = @_;

    if (ref $v eq 'HASH') {
        while (my ($k, $v) = each %$v) {
            local $COLLAPSE_KEY = length($COLLAPSE_KEY) ? sprintf('%s[%s]', $COLLAPSE_KEY, $k) : $k;
            _collapse($v, $r);
        }
    } elsif (ref $v eq 'ARRAY') {
        for (@$v) {
            local $COLLAPSE_KEY = $COLLAPSE_KEY . '[]';
            _collapse($_, $r);
        }
    } elsif (ref $v eq 'JSON::XS::Boolean' || ref $v eq 'JSON::PP::Boolean') { # could use JSON::is_bool, but that would require extra dependency
        $r->{$COLLAPSE_KEY} = $v;
    } elsif (!ref $v) {
        $r->{$COLLAPSE_KEY} = $v;
    } else {
        my $ref = ref $v;
        Carp::confess("${ref} is not supported by collapse_nested_params");
    }
}

sub collapse_nested_params {
    my $dat = shift;
    if (ref $dat ne 'HASH') {
        Carp::croak("The argument should be HashRef");
    }

    local $COLLAPSE_KEY = '';
    my $r = +{};
    _collapse($dat, $r);
    return $r;
}

1;
__END__

=encoding utf-8

=head1 NAME

Data::NestedParams - entry[title]=foo&tags[]=art&tags[]=modern

=head1 SYNOPSIS

    use Data::NestedParams;

    my $expanded = expand_nested_params(
        [
            'entry[title]' => 'foo',
            'tags[]' => 'art',
            'tags[]' => 'modern',
        ]
    );
    # $expanded = { entry => {title => 'foo'}, tags => ['art', 'modern'] };

=head1 DESCRIPTION

Ruby on Rails has a nice feature to create nested parameters that help with the organization of data in a form - parameters can be an arbitrarily deep nested structure.

The way this structure is denoted is that when you construct a form the field names have a special syntax which is parsed.

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom@gmail.comE<gt>

=head1 SEE ALSO

L<Catalyst::Plugin::Params::Nested>, L<CGI::Expand>
L<https://github.com/rack/rack/blob/master/lib/rack/utils.rb#L90>

=cut

